h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/task/XGBoostUploadMatrixTask.java

```diff
@@ -15,17 +15,23 @@
     private final FrameMatrixLoader matrixLoader;
     private final String[] remoteNodes;
     private final boolean https;
+    private final String userName;
+    private final String password;
 
-    public XGBoostUploadMatrixTask(Key modelKey, boolean[] frameNodes, FrameMatrixLoader loader, String[] remoteNodes, boolean https) {
+    public XGBoostUploadMatrixTask(
+        Key modelKey, boolean[] frameNodes, FrameMatrixLoader loader, String[] remoteNodes, boolean https, String userName, String password
+    ) {
         super(modelKey, frameNodes);
         this.matrixLoader = loader;
         this.remoteNodes = remoteNodes;
         this.https = https;
+        this.userName = userName;
+        this.password = password;
     }
 
     @Override
     protected void execute() {
-        XGBoostHttpClient client = new XGBoostHttpClient(remoteNodes[H2O.SELF.index()], https);
+        XGBoostHttpClient client = new XGBoostHttpClient(remoteNodes[H2O.SELF.index()], https, userName, password);
```

Comment:
New XGBoostHttpClient is created every single time. By looking at it's constructor, it looks kind-of expensive to construct.

- Pool it (with destruction timeout ? Ideal scenario would be trigger deletion after model training completion) - preferred option.
- Make it cheaper to instantiate by lazily instantiating it once.

Reply:
yes, but this is done only once per node, only on leader node this is done twice per training

Reply:
That's good then.